### PR TITLE
Adding function step

### LINKED ISSUE #1295: Make a function step
Similar to the Unix step, make a function step that is similar to the function ensemble that was created in https://github.com/equinor/ert/tree/ee-function-ensemble.

### PR DESCRIPTION
**Pull request by DanSava**
**Issue**
Resolves #1295 

**Approach**
Add a function step class that allows the execution of simple python code

### CODE REVIEW

File: ert3/evaluator/poly.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import json
+
+def polynomial(a, b, c):
+    x_range = tuple(range(10))
+    result =  tuple(a * x**2 + b * x + c for x in x_range)
+    with open("output_function.json", "w") as f:
```
Comment (markusdregi) [inline ert3/evaluator/poly.py:6]: Why does the function write to a file?

File: examples/polynomial/evaluation/ensemble.yml [hunk 2]
```diff
@@ -8,4 +8,5 @@ input:
 forward_model:
   driver: local
   stages:
-    - evaluate_polynomial
+    - evaluate_polynomial_script
```
Comment (markusdregi) [inline examples/polynomial/evaluation/ensemble.yml:11]: Does this make any sense as an example?

**Review by markusdregi — COMMENTED** [review]
Could you write a description of the approach taken in this PR @DanSava? I'm sure there are some interesting design choices you had to make along the way :) Among others:

- is the function output to be persisted in a file?
- the example seems to indicate that one can run both function and unix steps, however it is not clear to me how one redirects output from one stage to another? If the later is not addressed, perhaps it is better to limit the PR to allow for a single step (that now can be a function step in addition to a unix step)?

File: ert3/evaluator/poly.py (hunk 1 above)
Comment (DanSava) [reply ert3/evaluator/poly.py:6]: Just for the sake of using the mechanism that loads results from files in storage. I can move the file writing outside the function. What do you think ?

File: examples/polynomial/evaluation/ensemble.yml (hunk 2 above)
Comment (DanSava) [reply examples/polynomial/evaluation/ensemble.yml:11]: Are you referring to having a script and a function calculating the same polynomial? I was aiming for an example that shows you can have more than one stage may be the naming of the stage is not very inspired but I am open to suggestions regarding how a better example would look like.

**Comment by DanSava** [comment]
> is the function output to be persisted in a file

For now, if we want to have results in storage the function output needs to be persisted to a file, when the storage integration will be done I was thinking the results would be persisted directly to storage without going through files. 

> * the example seems to indicate that one can run both function and unix steps, however it is not clear to me how one redirects output from one stage to another? If the later is not addressed, perhaps it is better to limit the PR to allow for a single step (that now can be a function step in addition to a unix step)?

Currently, the prefect implementation supports redirecting output from a step when it is defined as input in another step, so it does not really matter from which stage the steps are coming from as long as the input and output files are defined consistently.

File: ert3/evaluator/poly.py (hunk 1 above)
Comment (markusdregi) [reply ert3/evaluator/poly.py:6]: I think that would be better! If we need to persist something to a file as a temporary thing to make the wheels turn that should be hidden from our users and a plan (and issue) should be created for how to deal with this :)

File: examples/polynomial/evaluation/ensemble.yml (hunk 2 above)
Comment (markusdregi) [reply examples/polynomial/evaluation/ensemble.yml:11]: Yes, I was thinking that we should strive to keep the examples instructive. So if we have an example with only a function the above would be a good example (if the unix_step is removed). If we want to display the conjunction between two steps, then I think we should create an example that high lights this. If the later, I can try to think of a good example...

**Comment by markusdregi** [comment]
> if we want to have results in storage the function output needs to be persisted to a file

This I think we definitively want! Otherwise the step type will leak very far...

**Comment by markusdregi** [comment]
> the prefect implementation supports redirecting output from a step when it is defined as input in another step

Aha, so is it then so that if the output record name of one step matches the input record name of another step then the data will be forwarded from the first step to the other?

File: examples/polynomial/stages.yml [hunk 3]
```diff
@@ -18,3 +18,21 @@
 
   script:
     - poly --coefficients coefficients.json --output output.json
+
+
+-
+  name: evaluate_polynomial_function
+  environment: polynomial
```
Comment (markusdregi) [inline examples/polynomial/stages.yml:25]: The environment key have been removed from the configuration...

File: examples/polynomial/stages.yml [hunk 4]
```diff
@@ -18,3 +18,21 @@
...
+    -
+      name: polynomial
+      args: coefficients.json
+
+  output:
+    -
+      record: polynomial_function_output
+      location: output_function.json
+
+  script:
```
Comment (markusdregi) [inline examples/polynomial/stages.yml:37]: If a function step is to be a minimal chunk of work, do we need it to support multiple function calls? Should we consider letting a `function_step` be a single Python function and then if one wants to combine multiple functions that is done via the stages and the forward model?

Comment (DanSava) [reply examples/polynomial/stages.yml:37]: > Aha, so is it then so that if the output record name of one step matches the input record name of another step then the data will be forwarded from the first step to the other?

Yes, that's how it works now.

File: examples/polynomial/evaluation/ensemble.yml (hunk 2 above)
Comment (DanSava) [reply examples/polynomial/evaluation/ensemble.yml:11]: How about we keep one stage in the example and the stage will contain both a script and a function. But the question is what should the function do? 
Should it just calculate the polynomial one more time or should it calculate something else like summing the output of the script?

**Comment by markusdregi** [comment]
As mentioned on video I was hoping to be able to make a minimal function step along the lines of:

```yaml
-
  name: evaluate_polynomial_function
  step_type: function

  input:
    - record: coefficients

  output:
    - record: polynomial_output

  function: ert3.evaluator.poly:polynomial
```

We would then still only allow a single step, either a unix step or a function step, but then we can at least run a Python function and use the two types interchangeably from the perspective of the rest of the setup. And then we can postpone the discussion about syntax for passing data from one step to another within a "pure" stage and the forward model until later.

**Comment by markusdregi** [comment]
I'll think about it until tomorrow and then we can continue the discussion. Maybe even bring it up at the technical planning?

File: ert_shared/ensemble_evaluator/prefect_ensemble/function_step.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+        stage_id = self.get_stage_id()
+        step_id = self.get_step_id()
+        return f"/ert/ee/{self._ee_id}/real/{iens}/stage/{stage_id}/step/{step_id}"
+
+    def function_source(self, fun_id):
+        return f"{self.step_source}/job/{fun_id}"
+
+    def run_job(self, client, job):
+        try:
+            result = job["executable"](**self._step["step_input"])
```
Comment (mortalisk) [inline ert_shared/ensemble_evaluator/prefect_ensemble/function_step.py:54]: Are the step inputs supposed to be only literal values, or should they also be able to come from other tasks via storage?

Comment (DanSava) [reply ert_shared/ensemble_evaluator/prefect_ensemble/function_step.py:54]: They should be able to come from tasks as well via storage, but that is currently not really well defined and to just get things started I assume only literal values for now.

File: ert3/config/_stages_config.py [hunk 6]
```diff
@@ -1,9 +1,24 @@
 import importlib
 import os
-from typing import List, Callable, Optional
+from typing import List, Callable, Optional, Union
```
Comment (mortalisk) [inline ert3/config/_stages_config.py:3]: Union is not used

File: ert3/evaluator/_evaluator.py [hunk 7]
```diff
@@ -21,17 +21,14 @@ def _create_evaluator_tmp_dir(workspace_root, evaluation_name):
...
-    # The current implementation only support one stage as the forward model
-    # and hence we fail if multiple are provided
-    assert len(ensemble.forward_model.stages) == 1
-
-
-def _prepare_input(ee_config, stages_config, inputs, evaluation_tmp_dir):
+def _prepare_input(ee_config, ensemble, stages_config, inputs, evaluation_tmp_dir):
     tmp_input_folder = evaluation_tmp_dir / "prep_input_files"
-    os.makedirs(tmp_input_folder)
+    os.makedirs(tmp_input_folder, exist_ok=True)
```
Comment (markusdregi) [inline ert3/evaluator/_evaluator.py:26]: Is this change really needed?

**Review by markusdregi — COMMENTED** [review]
This is starting to look really good 👏

File: ert3/evaluator/_evaluator.py (hunk 7 above)
Comment (DanSava) [reply ert3/evaluator/_evaluator.py:26]: No, just forgot it, it is much easier to test with it running ert3 without the need to always remove the `tmp` folder

File: ert3/config/_stages_config.py [hunk 8]
```diff
@@ -35,27 +50,87 @@ def is_executable(cls, location):
...
+        return _import_from(function)
+
 
 class StagesConfig(BaseModel):
     __root__: List[Step]
 
     def step_from_key(self, key: str) -> Optional[Step]:
         return next((step for step in self if step.name == key), None)
 
+    def get_prefect_stage(self, stage_name) -> List[dict]:
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/config/_stages_config.py:94]: Sure that sounds good

File: ert3/evaluator/_evaluator.py [hunk 9]
```diff
@@ -42,48 +39,24 @@ def _prepare_input(ee_config, stages_config, inputs, evaluation_tmp_dir):
     return input_files
 
 
+def _assert_single_stage_forward_model(stages_config, ensemble):
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/evaluator/_evaluator.py:42]: Definitely

File: ert_shared/ensemble_evaluator/prefect_ensemble/storage_driver.py [hunk 10]
```diff
@@ -27,6 +28,14 @@ def store(self, local_name, iens=None):
         shutil.copyfile(self._run_path / local_name, storage_uri)
         return storage_uri
 
+    def store_data(self, data, file_name, iens=None):
```
Comment (markusdregi) [inline ert_shared/ensemble_evaluator/prefect_ensemble/storage_driver.py:31]: This could potentially also be used by the `ert3.evaluator` as it currently writes the data to temp files and sends those in through the configuration...

File: examples/polynomial/stages.yml [hunk 11]
```diff
@@ -16,5 +16,21 @@
...
+    -
+      record: coefficients
+      location: coeffs
+
+  output:
+    -
+      record: polynomial_output
+      location: output
+
+  function: fun.poly:polynomial
```
Comment (markusdregi) [inline examples/polynomial/stages.yml:36]: 👏

File: examples/polynomial/fun/poly.py [hunk 12]
```diff
@@ -0,0 +1,7 @@
+def polynomial(coefficients: dict) -> tuple:
```
Comment (markusdregi) [inline examples/polynomial/fun/poly.py:1]: Should we name the module `function_steps` instead, just to make it clear what it actually contains?

File: tests/ert3/config/test_stages_config.py [hunk 13]
```diff
@@ -128,19 +148,78 @@ def test_step_unknown_script(tmpdir):
         ert3.config.load_stages_config(config)
 
 
-def test_stages_get_script():
+def test_step_function_errors():
```
Comment (markusdregi) [inline tests/ert3/config/test_stages_config.py:151]: Feel free to split this test up into multiple smaller tests...

File: tests/test_formatting.py [hunk 14]
```diff
@@ -24,8 +24,13 @@ def test_code_style(source_root):
     py_files_ert_gui = get_py_files(os.path.join(source_root, "ert_gui"))
     py_files_tests = get_py_files(os.path.join(source_root, "tests"))
     py_files_test_data = get_py_files(os.path.join(source_root, "test-data"))
+    py_files_ert3 = get_py_files(os.path.join(source_root, "ert3"))
```
Comment (markusdregi) [inline tests/test_formatting.py:27]: Hmm, I was actually under the impression that one had to explicitly opt out if one did not want to be style checked. But apparently I remember it wrong!

File: ert3/config/_stages_config.py [hunk 15]
```diff
@@ -35,20 +52,40 @@ def is_executable(cls, location):
...
-    script: List[str]
+    type: str
+    script: Optional[List[str]] = []
     input: List[InputRecord]
     output: List[OutputRecord]
-    transportable_commands: List[TransportableCommand]
+    transportable_commands: Optional[List[TransportableCommand]] = []
+    function: Optional[Callable]
+
+    def output_locations(self):
```
Comment (markusdregi) [inline ert3/config/_stages_config.py:62]: I think these support functions are quite tailored towards the currently expected format of the `PrefectEvaluator`. My suggestion would be to move this data fetching into the `ert3.evaluator` module instead...

**Review by markusdregi — COMMENTED** [review]
Had a few more comments :)

File: ert3/config/_stages_config.py [hunk 16]
```diff
@@ -1,9 +1,26 @@
...
 from typing import List, Callable, Optional
 from pydantic import root_validator, validator, FilePath, BaseModel
 
 
+def _import_from(path):
+    if ":" not in path:
+        raise ValueError("Function should be defined as module:function")
+    module, func = path.split(":")
+    try:
+        sys.path.append(os.getcwd())
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/config/_stages_config.py:13]: It can be removed after loading the module but that is not needed it will only be part of sys paths only during the runtime of the application.... But now I remember the discussion regarding possibly polluting the module namespace. 

I will remove the path after loading the module.

File: ert3/config/_stages_config.py [hunk 17]
```diff
@@ -35,20 +52,40 @@ def is_executable(cls, location):
 
 class Step(_StagesConfig):
     name: str
-    script: List[str]
+    type: str
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/config/_stages_config.py:55]: For now yes there are only two options, but I don't expect a lot more in the near future so a `Literal` is a good suggestion. Thanks.

File: ert_shared/ensemble_evaluator/prefect_ensemble/storage_driver.py (hunk 10 above)
Comment (DanSava) [reply ert_shared/ensemble_evaluator/prefect_ensemble/storage_driver.py:31]: Sure it can also be used in `ert3.evaluator` I did not add it there just to keep changes to a minimum.

File: ert_shared/ensemble_evaluator/prefect_ensemble/client.py [hunk 18]
```diff
@@ -1,7 +1,7 @@
 import websockets
 from websockets import ConnectionClosedOK
-
 import asyncio
+from cloudevents.http import to_json, CloudEvent
```
[the remark this answers is not in the document]
Comment (DanSava) [reply ert_shared/ensemble_evaluator/prefect_ensemble/client.py:4]: I think it is a matter of preference I also think that `cloudevents.http.to_json` would be more informative when reading the code but with a decent IDE I don't find it is vital.

File: ert_shared/ensemble_evaluator/prefect_ensemble/storage_driver.py (hunk 10 above)
Comment (ertomatic) [reply ert_shared/ensemble_evaluator/prefect_ensemble/storage_driver.py:31]: That is fair! I was just thinking it would clean up the code... And these things are always easier to do while you are at it. But, you can also make an issue or a follow-up PR branching out from this one :)

File: ert_shared/ensemble_evaluator/prefect_ensemble/client.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (ertomatic) [reply ert_shared/ensemble_evaluator/prefect_ensemble/client.py:4]: Touché! I’m still not leaving vim though 😂

File: ert3/config/_stages_config.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (ertomatic) [reply ert3/config/_stages_config.py:13]: I was just thinking that since cloudevent is all crazy, we might end up with elements on the path on the compute node that is not available on disk anymore or similar odd situations 🤷

**Comment by markusdregi** [comment]
Review bots are moving at an incredible pace 🤖

File: ert3/config/_stages_config.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/config/_stages_config.py:55]: It seems that:
https://pydantic-docs.helpmanual.io/usage/types/#literal-type 
Literal Type is a new feature of the python standard library as of python 3.8; prior to python 3.8, it requires the typing-extensions package.

So it does not really work for py36 without typing-extensions (https://pypi.org/project/typing-extensions/) I would like to avoid adding it as a dependency with this PR

[the remark this answers is not in the document]
Comment (markusdregi) [reply ert3/config/_stages_config.py:55]: It is already in use in the experiment config. So you will not have to do more than include it :)

File: ert3/config/_stages_config.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (markusdregi) [reply ert3/config/_stages_config.py:13]: Also, the `os.getcwd()` breaks the current social promise that any `ert3` command should succeed from any place within the workspace. However, if we want to keep this, we should make tests for it. Perhaps you could make an issue and point to this line to at least highlight the discussion?

[the remark this answers is not in the document]
Comment (DanSava) [reply ert3/config/_stages_config.py]: Not really sure what "social promise" you are referring to as far as I can tell `ert3 run` in the current example only works when running from the workspace root, but thanks for pointing it out I fixed the PR so when evaluation a function it will work from anywhere in the workspace

[the remark this answers is not in the document]
Comment (markusdregi) [reply ert3/config/_stages_config.py]: You are right! It used to work though, and everything outside loading the configuration should be tooling on top of the `workspace_root`, which is identified by moving up in the directory structure looking for `.ert`. But, the parsing of the configuration files are currently not done relative to the file.

The reason I named it a "social promise" was that it was an idea at some point, but as it is neither documented nor tested I have a hard time calling it a feature :)

File: ert3/config/_stages_config.py [hunk 19]
```diff
@@ -1,9 +1,32 @@
...
 
 
+def _import_from(path):
+    if ":" not in path:
+        raise ValueError("Function should be defined as module:function")
+    module, func = path.split(":")
+    try:
+        module = importlib.import_module(module)
+    except ModuleNotFoundError as err:
+        raise ValueError(f"No module named: {module}") from err
```
Comment (jondequinor) [inline ert3/config/_stages_config.py:22]: Why "degrade" this exception to `ValueError`? Why catch it at all? Isn't the `ModuleNotFound` _exactly_ right?

File: ert3/config/_stages_config.py [hunk 20]
```diff
@@ -1,9 +1,32 @@
...
+    try:
+        func = getattr(module, func)
+    except AttributeError as err:
+        raise ValueError(f"No function named: {func}") from err
```
Comment (jondequinor) [inline ert3/config/_stages_config.py:26]: Could be an `ImportError(name=func, path=module)`

File: ert3/console/_console.py [hunk 21]
```diff
@@ -139,7 +139,10 @@ def _load_ensemble_config(workspace, experiment_name):
 
 def _load_stages_config(workspace):
     with open(workspace / "stages.yml") as f:
-        return ert3.config.load_stages_config(yaml.safe_load(f))
+        sys.path.append(str(workspace))
+        config = ert3.config.load_stages_config(yaml.safe_load(f))
+        # sys.path.remove(str(workspace))
```
Comment (jondequinor) [inline ert3/console/_console.py:144]:
```suggestion

```

File: ert_shared/ensemble_evaluator/prefect_ensemble/prefect_ensemble.py [hunk 22]
```diff
@@ -199,33 +201,50 @@ def get_ordering(self, iens):
...
+            stage_id=step["stage_id"],
+            ee_id=ee_id,
+            on_failure=partial(self._on_task_failure, url=dispatch_url),
+            run_path=self.config.get("run_path"),
+            storage_config=self.config.get("storage"),
+            max_retries=self.config.get("max_retries", 0),
+            retry_delay=timedelta(seconds=2)
+            if self.config.get("max_retries") > 0
+            else None,
+        )
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/prefect_ensemble/prefect_ensemble.py:235]: These magic strings have a short shelf life. Please use non-magic strings.

File: ert_shared/ensemble_evaluator/prefect_ensemble/unix_step.py [hunk 23]
```diff
@@ -65,9 +65,35 @@ def storage_driver(self, run_path):
...
+            event = CloudEvent(
+                {
+                    "type": ids.EVTYPE_FM_JOB_FAILURE,
+                    "source": f"/ert/ee/{self._ee_id}/real/{self._iens}/stage/{self._stage_id}/step/{self._step_id}/job/{job['id']}",
+                    "datacontenttype": "application/json",
+                },
+                {ids.STDERR: cmd_exec.stderr, ids.STDOUT: cmd_exec.stdout},
+            )
+            client.send(to_json(event).decode())
+            raise RuntimeError(
```
Comment (jondequinor) [inline ert_shared/ensemble_evaluator/prefect_ensemble/unix_step.py:90]: `OSError` could fit well here.

File: ert3/evaluator/_evaluator.py [hunk 24]
```diff
@@ -54,38 +52,54 @@ def _build_ee_config(evaluation_tmp_dir, ensemble, stages_config, input_records)
...
+                "name": stage.function.__name__,
+                "executable": stage.function,
+                "output": output_locations[0],
+            }
+        )
 
-    command_scripts = [cmd.location for cmd in stage.transportable_commands]
-    output_files = [out.location for out in stage.output]
+    for script in stage.script:
+        name, *args = script.split()
```
Comment (jondequinor) [inline ert3/evaluator/_evaluator.py:76]: Do we want to do this ourselves? 

I'm just thinking if you want to run this command `ecl_run "C:\My Documents\My First Eclipse Simulation" …` you're screwed.

Anything we can lean on? E.g. something from argparse to get the actual constituents of the command?

**Review by jondequinor — CHANGES_REQUESTED** [review]
As more and more humans are working on EE/ert3, I think magic strings becomes an issue—fast. Could you replace them with constants?

**Comment by markusdregi** [comment]
I've reviewed the functionality provided to the user in this PR and I think it looks very good 👍 I appreciate the evaluator based test and the configuration test. In addition it would be good if we could have an end to end integration test of the functionality. My suggestion would be to add a run of the new experiment in the polynomial workspace to the `run_demo` script, as this will cause the experiment to the executed during CI.

In addition I think the responsibilities of the feature is spread out over the `ert3` module according to the intentions of the different modules.

I have however not looked into the detailed implementation of each function, exceptions raised and checked whether the configuration tests cover the non-happy combinations of being something in between a `unix` and a `function` step.

Also, my inspection of the implementation intersecting with the `prefect_evaluator`, including the tests is very shallow.

I hope you find someone to continue the review process during next week. It seems to me that @jondequinor is already of to a good start 👏 The best of luck 😎

File: ert3/evaluator/_evaluator.py (hunk 24 above)
Comment (DanSava) [reply ert3/evaluator/_evaluator.py:76]: I share your concern but I am not really changing the behaviour with this PR and I would do any improvements in future improvements PR. 

But I will look if there is anything from argparse that can help.

**Comment by DanSava** [comment]
> Also, my inspection of the implementation intersecting with the `prefect_evaluator`, including the tests is very shallow.

Not really sure what this is referring to.

**Review by jondequinor — APPROVED** [review]
LGTM!

**Comment by sondreso** [comment]
Jenkins test this please
